bin/main

```diff
@@ -0,0 +1,137 @@
+#!/usr/bin/env ruby
+
+# frozen_string_literal: true
+
+module Enumerable
+  def my_each
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+    i = 0
+    while i < size
+      yield(self[i])
+      i += 1
+    end
+    entry
+  end
+
+  def my_each_with_index
+    return to_enum unless block_given?
+
+    entry = is_a?(Range) ? to_a : self
+
+    i = 0
+    while i < size
+      yield(self[i], i)
+      i += 1
+    end
+    entry
+  end
+
+  def my_select
+    return to_enum unless block_given?
+
+    i = 0
+    arr = []
+    while i < size
+      arr << self[i] if yield(self[i])
+      i += 1
+    end
+    arr
+  end
+
+  def my_all?
+    return to_enum unless block_given?
+
+    elem = true
+    i = 0
+    while i < size
+      elem = false unless yield(self[i])
+      i += 1
+    end
+    elem
+  end
```

Comment:
* [ ] my_all? when no block or argument is given returns true when none of the collection members are false or nil
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all? == array.all? # false
  ```
* [ ] my_all? when a class is passed as an argument returns true if all of the collection is a member of such class
  ```ruby
    array = [1,2,5,4,7]
    p array.my_all?(Integer) == array.all?(Integer) # false
  ```
* [ ] my_all? when a Regex is passed as an argument returns true if all of the collection matches the Regex
  ```ruby
    array = %w[dog door rod blade]
    p array.my_all?(/c/) == array.all?(/c/) # false
  ```
* [ ] my_all? when a pattern other than Regex or a Class is given returns true if all of the collection matches the pattern
  ```ruby
    array = %w[dog door rod blade]
    p array.my_all?(5) == array.all?(5) # false
  ```